### PR TITLE
fix the kmerSize bug. change the node.setReference to setCopy to solve t...

### PR DESCRIPTION
**Pull request by JavierJia**
...he read out of range problem

### CODE REVIEW

**Comment by JavierJia** [comment]
@jakebiesinger , you can merge this branch to build the graph.

File: genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/dataflow/KmerNodePairSequenceWriterFactory.java [hunk 1]
```diff
@@ -41,13 +41,13 @@
...
         this.confFactory = new ConfFactory(conf);
-        Kmer.setGlobalKmerLength(Integer.parseInt(conf.get(GenomixJobConf.KMER_LENGTH)));
     }
 
     public class TupleWriter implements ITupleWriter {
 
-        public TupleWriter(ConfFactory confFactory) {
+        public TupleWriter(ConfFactory confFactory) throws NumberFormatException, HyracksDataException {
             this.cf = confFactory;
+            Kmer.setGlobalKmerLength(Integer.parseInt(cf.getConf().get(GenomixJobConf.KMER_LENGTH)));
```
Comment (jakebiesinger) [inline genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/dataflow/KmerNodePairSequenceWriterFactory.java]: Can you update these setGlobalKmerLength references to the more general `GenomixJobConf.setGlobalStaticConstants(cf.getConf());`? That function sets the kmer length as well as any other "global constants"-- we don't have any others right now, but I think it's better practice.

File: genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/dataflow/ReadsKeyValueParserFactory.java [hunk 2]
```diff
@@ -62,6 +64,8 @@ public ReadsKeyValueParserFactory() {
         final FrameTupleAppender outputAppender = new FrameTupleAppender(ctx.getFrameSize());
         outputAppender.reset(outputBuffer, true);
 
+        Kmer.setGlobalKmerLength(kmerLength);
```
Comment (jakebiesinger) [inline genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/dataflow/ReadsKeyValueParserFactory.java]: Ditto.  Use GenomixJobConf.setGlobalStaticConstants.

File: genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/job/JobGen.java [hunk 3]
```diff
@@ -95,8 +94,7 @@ public JobGen(GenomixJobConf job, Scheduler scheduler, final Map<String, NodeCon
 
     protected void initGenomixConfiguration() throws HyracksDataException {
         Configuration conf = hadoopJobConfFactory.getConf();
-        kmerSize = Integer.parseInt(conf.get(GenomixJobConf.KMER_LENGTH));
-        Kmer.setGlobalKmerLength(kmerSize);
+        Kmer.setGlobalKmerLength(Integer.parseInt(conf.get(GenomixJobConf.KMER_LENGTH)));
```
Comment (jakebiesinger) [inline genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/job/JobGen.java]: and here.

**Comment by jakebiesinger** [comment]
Looks like it's working on HPC but the timing is a bit slow-- BUILD_HADOOP took 7.3min to run; BUILD_HYRACKS took 13.1min this round (compared to 9.9min before our recent genomix-data refactor).

File: genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/dataflow/KmerNodePairSequenceWriterFactory.java (hunk 1 above)
Comment (JavierJia) [inline genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/dataflow/KmerNodePairSequenceWriterFactory.java]: Done

File: genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/dataflow/ReadsKeyValueParserFactory.java (hunk 2 above)
Comment (JavierJia) [inline genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/dataflow/ReadsKeyValueParserFactory.java]: Done

File: genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/job/JobGen.java (hunk 3 above)
Comment (JavierJia) [inline genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/job/JobGen.java]: Done

Comment (jakebiesinger) [inline genomix/genomix-hyracks/src/main/java/edu/uci/ics/genomix/hyracks/graph/job/JobGen.java]: Seems like you missed this one?

**Comment by JavierJia** [comment]
The main change is the TreeSet stuff, isn't it ?

**Comment by jakebiesinger** [comment]
TreeSet changes, including a few data structures that are materialized at set-as-copy/reference time rather than when in use.  In other words, we used to keep more stuff as byte arrays until you used them.

But timing on BUILD_HADOOP hasn't changed much.  Previously, it was 7.75min and it's now 7.3min in one run and 7.5min on a redo.

In any case, I'm going to go ahead and merge this.

**Comment by jakebiesinger** [comment]
~~Hmm.. I guess I should have waited to commit until you fixed the one you missed...~~

Nevermind. Looks like it got in before the merge occurred...

**Pull request merged**